### PR TITLE
Keep example_inputs when saving and loading ExportedProgram

### PR DESCRIPTION
**Pull request by jwfromm**
Summary:
`torch.export` is a powerful tool for creating a structured and shareable package from arbitrary pytorch code. One great use case of `torch.export` is sharing models or subgraphs in a way that allows results to be easily replicated. However, in the current implementation of `export`, the `example_inputs` field is thrown out. When trying to replicate bugs, benchmarks, or behaviors, losing the original input shapes and values makes the process much messier.

This change adds saving and loading for the `example_inputs` attribute of an `ExportedProgram` when using `torch.export.save` and `torch.export.load`. This simple addition makes `ExportedPrograms`s a fantastic tool for performance and accuracy replication. For example, with this change we enable the following workflow:

```
# Script to create a reproducible accuracy issue with my model.
kwargs = {"fastmath_mode": True}
exp_program = export(my_model, sample_inputs, kwargs)
result = exp_program.module()(*sample_inputs, **kwargs)
# Uhoh, I dont like that result, lets send the module to a colleague to take a look.
torch.export.save(exp_program, "my_model.pt2")
```

My colleague can then easily reproduce my results llike so:

```
# Script to load and reproduce results from a saved ExportedProgram.
loaded_program = torch.export.load("my_model.pt2")
# The following line is enabled by this Diff, we pull out the arguments
# and options that caused the issue.
args, kwargs = loaded_program.example_inputs
reproduced_result = loaded_program.module()(*args, **kwargs)
# Oh I see what happened here, lets fix it.
```

Being able to share exact inputs and arguments makes `ExportedPrograms` much
more clean and powerful with little downside. The main potential issue with this change
is that it does slightly increase the size of saved programs. However, the size of
inputs will be much smaller than parameters in most cases. I am curious to hear
discussion on saved file size though.

The deserialization of `example_inputs` is currently implemented as `Optional`. Although this wont effect users of `export.save` and `export.load`, it does give backwards compatibility to any direct users of `serialize` and `deserialize`.

Test Plan:
This diff includes a new test which exercises the save / load flow with multiple args and kwargs.

```
buck test //caffe2/test:test_export -- TestSerialize
```

Differential Revision: D55294614

### CODE REVIEW

File: torch/_export/serde/serialize.py [hunk 1]
```diff
@@ -297,15 +299,15 @@ def serialize_torch_artifact(artifact: Dict[str, Any]) -> bytes:
         del copyreg.dispatch_table[FakeTensor]
 
 
-def deserialize_torch_artifact(serialized: Union[Dict[str, Any], bytes]):
-    if isinstance(serialized, dict):
+def deserialize_torch_artifact(serialized: Union[Dict[str, Any], Tuple[Any], bytes]):
```
Comment (zhxchen17) [inline torch/_export/serde/serialize.py:302]:
```suggestion
def deserialize_torch_artifact(serialized: Union[Dict[str, Any], Tuple[Any, ...], bytes]):
```

File: torch/_export/serde/serialize.py [hunk 2]
```diff
@@ -1287,6 +1290,7 @@ class Result:
         names_to_symbols: Dict[str, sympy.Symbol]
         state_dict: Dict[str, Union[torch.Tensor, torch.nn.Parameter]]
         constants: Dict[str, Union[torch.Tensor, torch.ScriptObject]]
+        example_inputs: Optional[Tuple[Tuple[torch.Tensor], Dict[str, Any]]]
```
Comment (zhxchen17) [inline torch/_export/serde/serialize.py:1293]:
```suggestion
        example_inputs: Optional[Tuple[Tuple[torch.Tensor, ...], Dict[str, Any]]]
```

File: torch/_export/serde/serialize.py [hunk 3]
```diff
@@ -1670,6 +1674,7 @@ def deserialize(
         serialized_graph_module: GraphModule,
         serialized_state_dict: Union[Dict[str, torch.Tensor], bytes],
         constants: Union[Dict[str, Any], bytes],
+        example_inputs: Optional[Union[Tuple[Tuple[torch.Tensor], Dict[str, Any]], bytes]]= None,
```
Comment (zhxchen17) [inline torch/_export/serde/serialize.py:1677]:
```suggestion
        example_inputs: Optional[Union[Tuple[Tuple[torch.Tensor, ...], Dict[str, Any]], bytes]]= None,
```

**Review by zhxchen17 — APPROVED** [review]
@jwfromm thanks for adding example inputs serialization! some nits on type annotations. (Tuple[Tensor] is something like (tensor,), and Tuple[Tensor, ...] is really for tuples with arbitrary length)

**Comment by jwfromm** [comment]
Thank you for the review @zhxchen17! I've updated the type annotations on your feedback and it looks like all (but one) tests are passing. The failing test appears to just be flaky and is unrelated to this change. What do you recommend for proceeding? Should I retrigger all testing or is there a more finegrained way to retrigger just the failing test?
